tokio/src/util/wake_list.rs

```diff
@@ -0,0 +1,44 @@
+use core::mem::{self, MaybeUninit};
+use std::task::Waker;
+
+const NUM_WAKERS: usize = 32;
+
+pub(crate) struct WakeList {
+    inner: [MaybeUninit<Waker>; NUM_WAKERS],
+    curr: usize,
+}
+
+impl WakeList {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: unsafe { MaybeUninit::uninit().assume_init() },
+            curr: 0,
+        }
+    }
+
+    pub(crate) fn can_push(&self) -> bool {
+        self.curr < NUM_WAKERS - 1
+    }
+
+    pub(crate) fn push(&mut self, val: Waker) {
+        debug_assert!(self.can_push());
+
+        self.inner[self.curr] = MaybeUninit::new(val);
+        self.curr += 1;
+    }
+
+    pub(crate) fn wake_all(&mut self) {
+        for waker in &mut self.inner[..self.curr] {
+            unsafe { mem::replace(waker, MaybeUninit::uninit()).assume_init() }.wake()
+        }
+        self.curr = 0;
+    }
+}
+
+impl Drop for WakeList {
+    fn drop(&mut self) {
+        for waker in &mut self.inner[..self.curr] {
+            mem::drop(unsafe { mem::replace(waker, MaybeUninit::uninit()).assume_init() });
+        }
```

Comment:
```suggestion
        let slice = std::ptr::slice_from_raw_parts_mut(self.inner.as_mut_ptr() as *mut Waker, self.curr);
        unsafe { std::ptr::drop_in_place(slice) };
```

Reply:
Will `drop_in_place` run `Drop` on all elements of the slice? It seems like this code should run Drop on the slice, which doesn't have Drop on it.

Reply:
This works because `[T]` does in fact have a destructor. The type that doesn't is `&mut [T]`. This is also how the `Vec` destructor works.